### ISSUE TITLE
 Failure: cannot instantiate ctype 'EVP_MD_CTX' of unknown size.

### ISSUE THREAD

**Issue by okulev** [google/oauth2client#641]
I reported this issue in https://code.google.com/p/google-cloud-sdk/issues/detail?id=1105 and I was redirected here.

When I upgraded `python-cryptography` Debian package from version 1.4-2 to version 1.5-1, the above error occurs (`Failure: cannot instantiate ctype 'EVP_MD_CTX' of unknown size.` ). Downgrading python-cryptography to 1.4 resolves the issue. So there is problem either in python-cryptography or in oauth2client.

Result from `python -c "import OpenSSL; print(OpenSSL.__version__)"` is `16.0.0`.

Python stack trace follows:

```
DEBUG: Exception stack trace:
    Traceback (most recent call last):
      File "/usr/lib/google-cloud-sdk/platform/gsutil/gslib/__main__.py", line 577, in _RunNamedCommandAndHandleExceptions
        collect_analytics=True)
      File "/usr/lib/google-cloud-sdk/platform/gsutil/gslib/command_runner.py", line 299, in RunNamedCommand
        return_code = command_inst.RunCommand()
      File "/usr/lib/google-cloud-sdk/platform/gsutil/gslib/commands/ls.py", line 446, in RunCommand
        bucket_fields=bucket_fields):
      File "/usr/lib/google-cloud-sdk/platform/gsutil/gslib/wildcard_iterator.py", line 456, in IterBuckets
        for blr in self._ExpandBucketWildcards(bucket_fields=bucket_fields):
      File "/usr/lib/google-cloud-sdk/platform/gsutil/gslib/wildcard_iterator.py", line 375, in _ExpandBucketWildcards
        provider=self.wildcard_url.scheme):
      File "/usr/lib/google-cloud-sdk/platform/gsutil/gslib/gcs_json_api.py", line 519, in ListBuckets
        global_params=global_params)
      File "/usr/lib/google-cloud-sdk/platform/gsutil/gslib/third_party/storage_apitools/storage_v1_client.py", line 423, in List
        config, request, global_params=global_params)
      File "/usr/lib/google-cloud-sdk/platform/gsutil/third_party/apitools/apitools/base/py/base_api.py", line 701, in _RunMethod
        http, http_request, **opts)
      File "/usr/lib/google-cloud-sdk/platform/gsutil/third_party/apitools/apitools/base/py/http_wrapper.py", line 351, in MakeRequest
        max_retry_wait, total_wait_sec))
      File "/usr/lib/google-cloud-sdk/platform/gsutil/gslib/util.py", line 1596, in WarnAfterManyRetriesHandler
        http_wrapper.HandleExceptionsAndRebuildHttpConnections(retry_args)
      File "/usr/lib/google-cloud-sdk/platform/gsutil/third_party/apitools/apitools/base/py/http_wrapper.py", line 341, in MakeRequest
        check_response_func=check_response_func)
      File "/usr/lib/google-cloud-sdk/platform/gsutil/third_party/apitools/apitools/base/py/http_wrapper.py", line 391, in _MakeRequestNoRetry
        redirections=redirections, connection_type=connection_type)
      File "/usr/lib/google-cloud-sdk/platform/gsutil/third_party/oauth2client/oauth2client/client.py", line 616, in new_request
        self._refresh(request_orig)
      File "/usr/lib/google-cloud-sdk/platform/gsutil/third_party/oauth2client/oauth2client/client.py", line 885, in _refresh
        self._do_refresh_request(http_request)
      File "/usr/lib/google-cloud-sdk/platform/gsutil/third_party/oauth2client/oauth2client/client.py", line 900, in _do_refresh_request
        body = self._generate_refresh_request_body()
      File "/usr/lib/google-cloud-sdk/platform/gsutil/third_party/oauth2client/oauth2client/client.py", line 1611, in _generate_refresh_request_body
        assertion = self._generate_assertion()
      File "/usr/lib/google-cloud-sdk/platform/gsutil/third_party/oauth2client/oauth2client/service_account.py", line 396, in _generate_assertion
        key_id=self._private_key_id)
      File "/usr/lib/google-cloud-sdk/platform/gsutil/third_party/oauth2client/oauth2client/crypt.py", line 97, in make_signed_jwt
        signature = signer.sign(signing_input)
      File "/usr/lib/google-cloud-sdk/platform/gsutil/third_party/oauth2client/oauth2client/_openssl_crypt.py", line 100, in sign
        return crypto.sign(self._key, message, 'sha256')
      File "/usr/lib/python2.7/dist-packages/OpenSSL/crypto.py", line 2563, in sign
        md_ctx = _ffi.new("EVP_MD_CTX*")
    TypeError: cannot instantiate ctype 'EVP_MD_CTX' of unknown size
```

**Comment by jonparrott** [comment]
This smells like not directly are issue. Can you try using a virtualenv and see if this still happens? I'm worried that the `OpenSSL` package in `dist-packages` (owned by debian) is causing the conflict.

**Comment by okulev** [comment]
While looking at the problem, I noticed that a new `python-openssl` Debian package was released for new version 16.1.0-1 while I was on 16.0.0-2. After upgrading to it, and also upgrading `python-cryptography` to latest version 1.5-1 (from 1.4-2), and upgrading `google-cloud-sdk` to latest 125.0.0-0, the problem disappeared.
